CenterForOpenScience/osf.io
# [QA Migration Prep Required] Feature/Token-Scope relationships [PLAT-1062]

## Purpose

Change how personal access tokens and scopes are related.  Currently, scopes are represented as space-delimited strings on tokens (in both the db and API).  Embosf is requesting scopes be represented as relationships on tokens.

❗️ Make sure staging/prod has public scopes in db
❗️ Local testing will fail when trying to update/delete tokens until CAS-side portion is complete.  

## Changes

- Added a migration to represent scopes as an m2m field on token instead of a CharField
- Pull scopes from the db instead of public_scopes db
- Represent scopes as relationships instead of attributes on tokens, always embedded
- Modify requests to create/update tokens - sending scopes under relationships.  No option to remove all scopes entirely.
- Modify requests legacy front end makes to API
- Allow requests v2.15 and earlier to update scopes with attributes still, while 2.15 and later, use relationships
- Token scopes API endpoint added

## QA Notes

- Thoroughly test creating/updating/deleting personal access tokens
- Pre-migration prep, create a handful of tokens with different scopes. verify after migration that same scopes are on tokens.
- API tests, verify requests with early versions and later versions both work

## API notes
### Creating a token, early versions
```json
POST http://localhost:8000/v2/tokens/
{
  "data": {
    "type": "tokens",
    "attributes": {
    	"name": "my scope",
    	"scopes": "osf.full_read osf.full_write"
    }
  }
}
```

### Creating a token, later versions
```json
POST http://localhost:8000/v2/tokens/?version=2.16
{
  "data": {
    "type": "tokens",
    "attributes": {
    	"name": "my scope"
    },
    "relationships": {
    	"scopes": {
    		"data": [{
    			"id": "osf.full_read",
    			"type": "scopes"
    		}
    		]	
    	}
    }
    
  }
}
```

## Documentation

Tokens/scopes currently undocumented.
## Side Effects



## Ticket

https://openscience.atlassian.net/browse/PLAT-1062
